### ISSUE TITLE
Create PoW mining validation contracts

### ISSUE THREAD

**Issue by maraoz** [OpenZeppelin/openzeppelin-solidity#727]
## 🎉 Description

Idea taken from @Shrugs' image, in turn taken from 0xBitcoin:
![1_w262dszig5yxzaf2zg5uwq](https://user-images.githubusercontent.com/287189/36033642-657e28bc-0d90-11e8-8e4d-76ad26d2ad8c.png)

- [ ] 🐛 This is a bug report.
- [x] 📈 This is a feature request.

I'd make it generic, i.e not tied to token minting, as in the image, and maybe then add a `MinedToken` that uses the generic mining validation contracts

**Comment by arisAlexis** [comment]
interesting! how would a miner interact with this contract? you need to pay gas for each submission?

**Comment by alcuadrado** [comment]
This implementation is vulnerable to front-running. But you can easily fix that with something like
`bytes32 digest = keccak256(lastNonce, nonce, msg.sender);`.

This way the proof of work is tied to whoever did it, and no front-running is possible.

**Comment by Shrugs** [comment]
Updated contract should be

```
contract MineableToken is MintableToken {
  public uint256 blockHeight = 0;
  public uint256 lastNonce = 0;

  function getDifficulty() public returns (uint256) {
    return <some-difficulty-function>
  }

  function mint(uint256 nonce) public returns (bool success) {
    bytes32 digest = keccak256(msg.sender, lastNonce, nonce);
    require(uint256(digest) < getDifficulty());

    lastNonce = nonce
    blockHeight += 1
    return super.mint(msg.sender, 1);
  }
}
```

Mining would look like

```
const coinbase = web3.eth.accounts[0]
const lastNonce = await mineableToken.lastNonce()
const difficulty = await mineableToken.getDifficulty()
const randomNonce = () => BigNumber.random(78).mul(10 ** 78)  // this probably works, idk
let isValid = false
let nonce
while (!isValid) {
  nonce = randomNonce()
  const digest = web3.utils.keccack256(coinbase, lastNonce, nonce)
  // ^ make sure the bytes workout here, idk
  isValid = digest < difficulty
}

await mineableToken.mint(nonce)
// yay
```

**Comment by AugustoL** [comment]
Super interesting !! want to see it in zeppelin

It can also have a `getReward()` function, that returns the reward of the block number mined, it can be deflationary.

Also it can have a fee per transfer, like take `1 wei` for each transfer and claim all the `feeBalance` every time a block is mined, this way when you reach a block reward of cero you still have the fee reward.

The difficulty can be calculated knowing only how much time it took to mine the last block ? smth like:
```
if (lastBlockTime < blockTime)
  increaseDifficulty( blockTime.sub(lastBlockTime) )
else
  decreaseDifficulty( lastBlockTime.sub(blockTime) )
```

**Comment by alcuadrado** [comment]
I like the idea of validating PoW in a smart contract, but I wouldn't couple its implementation to minting a token. 

1. While miners do PoW for earning cryptocurrencies, that's not the purpose of the PoW. Block rewards are just a way of incentivize people to secure a blockchain. If obtaining new tokens is the entire purpose of the PoW it would only model the distribution of the tokens. The same distribution probably be obtained in a more environmental friendly.

2. It inhibit other (IMO more interesting) uses of it. The first uses that come to my mind: making spam and sibyl attacks more expensive, and hopefully impractical.

**Comment by Shrugs** [comment]
Yeah totally agreed, glad you brought that up! 

So now we've got a `Mineable` interface that implements PoW checking and exports a `verifyProofOfWork(nonce, difficulty)` function.

Then `MineableToken is Mineable, MintableToken` and mint looks like

```
function mint(_to, uint256 nonce) external returns (bool success) {
  require(verifyProofOfWork(nonce, getDifficulty()));
  
  adjustDifficulty();
  return mint(msg.sender, getReward();
}
```

**Comment by nventuro** [comment]
>It inhibit other (IMO more interesting) uses of it. The first uses that come to my mind: making spam and sibyl attacks more expensive, and hopefully impractical.

I really liked the whole PoW idea, but had never come up with a use case that went beyond toy projects such as minting tokens. This, however, seems super neat.

**Comment by nventuro** [comment]
After discussing this further with @alcuadrado, we came to the conclusion that placing PoW requirements for prevent spam wouldn't really be a great solution, since it is not very expensive to purchase computing power (much) more powerful than the equivalent of running JavaScript inside a browser.

I think many of us like this idea because it sounds cool, but it lacks a real use-case. Closing until one comes up.
